x-pack/plugins/synthetics/server/synthetics_service/project_monitor_formatter.ts

```diff
@@ -377,54 +379,101 @@ export class ProjectMonitorFormatter {
 
   private handleStaleMonitors = async () => {
     try {
-      const staleMonitorsData = Object.values(this.staleMonitorsMap).filter(
+      const staleMonitorsList = Object.values(this.staleMonitorsMap).filter(
         (monitor) => monitor.stale === true
       );
 
-      for (const staleMonitor of staleMonitorsData) {
-        if (!this.keepStale) {
-          await this.deleteStaleMonitor({
-            monitorId: staleMonitor.savedObjectId,
-            journeyId: staleMonitor.journeyId,
+      const encryptedMonitors = await this.savedObjectsClient.bulkGet<SyntheticsMonitor>(
+        staleMonitorsList.map((staleMonitor) => ({
+          id: staleMonitor.savedObjectId,
+          type: syntheticsMonitorType,
+        }))
+      );
+
+      let monitors = encryptedMonitors.saved_objects;
+
+      const hasPrivateMonitor = monitors.some((monitor) =>
+        monitor.attributes.locations.some((location) => !location.isServiceManaged)
+      );
+
+      if (hasPrivateMonitor) {
+        const {
+          integrations: { writeIntegrationPolicies },
+        } = await this.server.fleet.authz.fromRequest(this.request);
+        if (!writeIntegrationPolicies) {
+          monitors = monitors.filter((monitor) => {
+            const hasPrivateLocation = monitor.attributes.locations.some(
+              (location) => !location.isServiceManaged
+            );
+            if (hasPrivateLocation) {
+              const journeyId = (monitor.attributes as MonitorFields)[ConfigKey.JOURNEY_ID]!;
+              const monitorName = (monitor.attributes as MonitorFields)[ConfigKey.NAME]!;
+              this.handleStreamingMessage({
+                message: `Monitor ${journeyId} could not be deleted`,
+              });
+              this.failedStaleMonitors.push({
+                id: journeyId,
+                reason: 'Failed to delete stale monitor',
+                details: `Unable to delete Synthetics package policy for monitor ${monitorName}. Fleet write permissions are needed to use Synthetics private locations.`,
+              });
+            }
+            return !hasPrivateLocation;
           });
-        } else {
-          this.staleMonitors.push(staleMonitor.journeyId);
-          return null;
         }
       }
-    } catch (e) {
-      this.server.logger.error(e);
-    }
-  };
 
-  private deleteStaleMonitor = async ({
-    monitorId,
-    journeyId,
-  }: {
-    monitorId: string;
-    journeyId: string;
-  }) => {
-    try {
-      await deleteMonitor({
-        savedObjectsClient: this.savedObjectsClient,
-        server: this.server,
-        monitorId,
-        syntheticsMonitorClient: this.syntheticsMonitorClient,
-        request: this.request,
-      });
-      this.deletedMonitors.push(journeyId);
-      this.handleStreamingMessage({ message: `Monitor ${journeyId} deleted successfully` });
+      const chunkSize = 100;
```

Comment:
I think the chunking could be separated out into a different helper. I think this function itself could be organized a bit more as there is a lot of nesting and it's a bit difficult to reason about.